src/lib.ml

```diff
@@ -949,17 +957,28 @@ module Compile = struct
         ~pps:t.info.pps
         ~kind:(Lib_deps_info.Kind.of_optional t.info.optional)
     in
+    let requires_link =
+      t.requires >>= closure_with_overlap_checks db ~linking:false
+    in
+    let requires_compile = lazy (
+      t.requires >>= fun direct_requires ->
```

Comment:
Why don't we return `t.requires` directly here?

Reply:
The old closure was returned in top sorted order. If I don't preserve this order, the compilation would fail.

Reply:
That surprises me, why does it fail?

Reply:
I can only reproduce the failure when using `threads` instead of `threads.posix`. I'm not exactly sure if threads are special here, here it is:

```
    ocamlopt src/.dune.objs/native/dune__Scheduler.{cmx,o} (exit 2)
(cd _build/default && /Users/rgrinberg/.opam/4.07.1/bin/ocamlopt.opt -w @a-4-29-40-41-42-44-45-48-58-59-60-40 -strict-sequence -strict-formats -short-paths -keep-locs -g -I src/.dune.objs/byte -I src/.dune.objs/native -I src/catapult/.catapult.objs/byte -I src/catapult/.catapult.objs/native -I src/dag/.dag.objs/byte -I src/dag/.dag.objs/native -I src/dune_lang/.dune_lang.objs/byte -I src/dune_lang/.dune_lang.objs/native -I src/fiber/.fiber.objs/byte -I src/fiber/.fiber.objs/native -I src/memo/.memo.objs/byte -I src/memo/.memo.objs/native -I src/ocaml-config/.ocaml_config.objs/byte -I src/ocaml-config/.ocaml_config.objs/native -I src/stdune/.stdune.objs/byte -I src/stdune/.stdune.objs/native -I src/which_program/.which_program.objs/byte -I src/which_program/.which_program.objs/native -I src/xdg/.xdg.objs/byte -I src/xdg/.xdg.objs/native -I vendor/opam-file-format/src/.opam_file_format.objs/byte -I vendor/opam-file-format/src/.opam_file_format.objs/native -I vendor/re/src/.dune_re.objs/byte -I vendor/re/src/.dune_re.objs/native -intf-suffix .ml -no-alias-deps -opaque -open Dune -o src/.dune.objs/native/dune__Scheduler.cmx -c -impl src/scheduler.pp.ml)
File "src/scheduler.ml", line 35, characters 10-16:
Error: Unbound module Thread
```

The above is compiled with using `t.direct_required`

Reply:
I thought using `threads` instead of `threads.posix` was an issue independently of this anyway?

Reply:
Oh yeah. It's indeed independent. Sorry, I'm still half asleep here. Will get rid of this cruft.

Reply:
coding before :coffee: :)